### PR TITLE
interfaces/builtin: modify shutdown interface to allow access to SetRebootParameter method

### PR DESCRIPTION
**Pull request by bugraaydogar**
The D-BUS API of `systemd-logind` [1] allows developers to set a reboot parameter when they would like to change the default behavior of the reboot commands. However, the shutdown interface does not allow to use `SetRebootParameter`. This PR aims to solve it by allowing access to `SetRebootParameter` method.

[1] https://www.freedesktop.org/software/systemd/man/org.freedesktop.login1.html

### CODE REVIEW

**Review by alexmurray — COMMENTED** [review]
So it looks like by allowing this method, the caller can specify that on next reboot to halt the system, restart the system, suspend the system, or to also kexec into a new kernel. kexec'ing seems quite a privileged operation so I wonder if perhaps this interface should now become super-privileged as a result?

@bugraaydogar can you confirm if my understanding of this method is correct?

**Comment by bugraaydogar** [comment]
Hi @alexmurray, Your understanding is correct. It is also possible to pass `RB_KEXEC` parameter which could load a previously loaded kernel with `kexec_load`. That being said, I think we eventually do not allow to use `kexec_load` as mentioned here => https://github.com/snapcore/snapd/blob/master/interfaces/seccomp/template.go#L32

Therefore, I think we already avoid this risk. So, there is no reason to make `shutdown` interface super-privileged. But, It might be possible that I overlook some details and if this is the case, I'm happy to follow your suggestion.

**Comment by alexmurray** [comment]
Ok, I think given that `kexec_load` is blocked by seccomp that this is helpful, but I still think it would make sense for it to be super-privileged nonetheless - although I don't think we can make this change easily since then existing snaps in the store that use `shutdown` (but do not auto-connect it) would need a snap declaration now.

@pedronis  can you comment please?

**Comment by pedronis** [comment]
@alexmurray I'm  a bit confused why seccomp is relevant? the request will go to an uncofined process no? so that is not under our seccomp template, am I missing something?

**Comment by alexmurray** [comment]
@pedronis my concern was that this change would allow a snap that now plugs `shutdown` to cause the machine to kexec into a new kernel - however, to do that, something would first have to call `kexec_load` to load this new kernel - this system call is blocked by the default seccomp filter which means the snap itself cannot call this. And so the snap is then not able to cause a new kernel to be kexec'd since it can't call `kexec_load` itself. Hence why this is relatively safe.

However, I was thinking that the act of being able to trigger a kexec into some new kernel (even if a snap can't specify the kernel itself via kexec_load) was a change in scope for this interface and hence whether it would be appropriate for this to become super privileged - but then this has other impacts as I outlined above, hence why I asked for your input on that.

**Comment by Roadmaster** [comment]
There are 208 snaps where a currently-released revision has a shutdown plug. Of those, 113 have NO explicit allow-auto-connection (so they will stop being able to use shutdown). The remaining 95 do have an explicit auto-connection setting, either a boolean "true" or an "on-store" constraint.

**Comment by bugraaydogar** [comment]
Hi @alexmurray, Hi @pedronis,

Considering that 113 snaps might be affected if the shutdown interface becomes super-privileged, what would be your advice moving forward? Maybe we could work with @Roadmaster to configure all of the snaps.

**Comment by alexmurray** [comment]
Sure, from my point of view, if the store team feel confident they can grant the required snap declarations then I would be happy to proceed with adding this additional permission to the shutdown interface and making it super-privileged.

**Comment by Roadmaster** [comment]
I would need to do a bit of preliminary digging to understand if we can safely update them in "bulk" - this would be my preference; we can do it manually but it's going to be a drag with so many snaps.

File: interfaces/builtin/shutdown.go
Comment (pedronis) [inline interfaces/builtin/shutdown.go:30]: this is not correct for an implicit interface like this one, see kernel_module_control.go for something appropriate for this

**Review by pedronis — CHANGES_REQUESTED** [review]
switch to superprivileged is not correct for the kind of interface

File: interfaces/builtin/shutdown.go
Comment (bugraaydogar) [reply interfaces/builtin/shutdown.go:30]: so should I revert my changes on `shutdownBaseDeclarationSlots` and define `shutdownBaseDeclaration` with the following content?

```
const shutdownBaseDeclaration = `
  shutdown:
    allow-installation: false
    deny-auto-connection: true
`
```

**Comment by bugraaydogar** [comment]
@pedronis I made the required changes to make the `shutdown` interface super-privileged. I think, this PR waits @Roadmaster 's input on bulk assertion configuration.

**Comment by shanepelletier** [comment]
> for clarity for snaps that are not auto-connected but have a plug we would want to change their snap-declaration rules to:
> {"shutdown": {"allow-installation": true, "deny-auto-connection": true}}

@pedronis I'm a little confused here; since we want snaps that are not currently auto-connected to become auto-connected, wouldn't the correct rule for us (i.e., the store team) to use be `{"shutdown": {"allow-installation": true, "allow-auto-connection": true}}`? It's entirely possible I'm misunderstanding your comment or how declarations work :slightly_smiling_face:.

**Comment by pedronis** [comment]
@shanepelletier 

> snaps that are not currently auto-connected to become auto-connected

to be clear we do not want that, and it seems from the latter part of your message that is undersood. We want snaps currently using this to continue being installable but we do not want to turn not auto-connecting ones into auto-connecting.

**Comment by Roadmaster** [comment]
The logic Shane implemented is:


 1. For snaps that did not have an explicit "allow-auto-connection" value in their plugs, either `true` or a store-scoped value, we just add `{"shutdown": {"allow-installation": true, "deny-auto-connection": true}}` to the global plugs rules (append, not replace so we don't touch any other rules). These snaps did not have a previous shutdown entry in plugs.
 
 2. For snaps that DO have something explicit in "allow-auto-connection" for `shutdown` in their plugs rules, let's call that $ALLOW_AUTO. We need to set the existing shutdown element so it looks like this: 

`{"shutdown": {"allow-installation": $ALLOW_AUTO, "allow-auto-connection: $ALLOW_AUTO}}`

Please let us know if that looks correct. If you prefer to read actual code let me know and we can share the Python script Shane created.

**Comment by pedronis** [comment]
@Roadmaster  the plan essentially looks correct for snaps with a shutdown plug but

> A possible gap here is that for #1, we are not store-scoping the newly-added rule. @pedronis should we do it? e.g. if a store has scoping turned on, and we do "allow-installation": true, things will break, right?

I think store scoping needs to be used instread of true if the snap is in a brand store with store scoping enabled, as otherwise some of the store consistency checks will fail at some point

**Comment by Roadmaster** [comment]
Thanks, Shane updated our code so for #1:

For snaps that did not have an explicit "allow-auto-connection" value in their plugs, either true or a store-scoped value, we just add `{"shutdown": {"allow-installation": $ALLOW_AUTO, "deny-auto-connection": true}}` to the global plugs rules (append, not replace so we don't touch any other rules). These snaps did not have a previous shutdown entry in plugs.

$ALLOW_AUTO will be either `true` or a store-scoped rule `"in-store": ["whatever"]` depending on whether scoping is enabled for the store the snap is in (this is the change we discussed earlier).

**Comment by Roadmaster** [comment]
@pedronis hey if you could check the latest comment and confirm whether this is ok we'd appreciate it. Once this is merged we can test on a couple of snaps to ensure the changes make sense and then go for the rest.

**Review by pedronis — APPROVED** [review]
+1

**Comment by pedronis** [comment]
@Roadmaster just a reminder that we can run the scripts at any time but we also have to re-run them after this change goes to stable and until the corresponding base-declaration changes are finally used by store reviews, to cover any snap that gets updated in that mean time. Ideally we would change the store behavior before snapd itself that means using a base-declaration from a beta snapd or something like that in review-tools cc @alexmurray

**Comment by alexmurray** [comment]
I can definitely merge the associated changes into review-tools - just let me know when this is likely to be required store-side.

**Comment by Roadmaster** [comment]
> I can definitely merge the associated changes into review-tools - just let me know when this is likely to be required store-side.

Hi Alex, we'd like to coordinate with you or someone from your team to move this forward.

Can we apply the change to a couple of snaps to ensure the result looks sane? The snap-declaration is accessible via curl so we can run this on our side for a few snaps and then give you URLs to check.

If so, would the following work:

 1. We test a couple of snaps. If they look OK we proceed.
 2. We apply the bulk change
 3. We deploy review-tools with the updated base declaration.
 4. We apply the bulk change again to maybe catch any stragglers or unverified snap updates.
 5. profit?

I'm just unsure of the order of steps 2 and 3. If we need to deploy review-tools first and assuming that doesn't have an impact on incoming snap uploads, we can swap them in order.

- Daniel

**Comment by alexmurray** [comment]
The order of 2 and 3 probably doesn't matter too much - but if we do deploy review-tools first and then take too long to update the various snap declarations in the store then we risk too many snaps getting stuck blocked and stuck in the review queue and needing manual approval.

But I like this approach - let's update the declaration on a few snaps first that already plug shutdown, then update and deploy review-tools, rerun the automated review to make sure the updated snaps pass as expected, then do the bulk update for the rest of the store.

**Comment by pedronis** [comment]
tentatively schedule this for 2.61

**Comment by pedronis** [comment]
@alexmurray the store has changed the preexisting snap-declarations

mmh, there's a conflict now

**Comment by pedronis** [comment]
I fixed the conflict

**Comment by bugraaydogar** [comment]
@mvo5 @pedronis, is there anything blocking the merge of this PR?

**Comment by pedronis** [comment]
> @mvo5 @pedronis, is there anything blocking the merge of this PR?

we need @alexmurray conclusive +1

**Review by alexmurray — APPROVED** [review]
Apologies, I thought I had already approved this - +1 from me!

**Pull request merged**
